taichi/codegen/codegen_opengl.cpp

```diff
@@ -167,156 +171,210 @@ class KernelGen : public IRVisitor {
+          "};\n";
     }
     if (used.global_temp) {
       kernel_header +=
           "layout(std430, binding = 2) buffer gtmp_i32 { int _gtmp_i32_[]; };\n"
           "layout(std430, binding = 2) buffer gtmp_f32 { float _gtmp_f32_[]; "
           "};\n"
           "layout(std430, binding = 2) buffer gtmp_f64 { double _gtmp_f64_[]; "
-          "};\n"
-          "#define _gtx_i32(x) _gtmp_i32_[(x) >> 2]\n"
-          "#define _gtx_f32(x) _gtmp_f32_[(x) >> 2]\n"
-          "#define _gtx_f64(x) _gtmp_f64_[(x) >> 3]\n";
+          "};\n";
     }
     if (used.extra_arg) {
       kernel_header +=
           "layout(std430, binding = 3) buffer earg_i32 { int _earg_i32_[]; "
-          "};\n" +
-          fmt::format("#define _extra_arg(i, j) _earg_i32_[(i) * {} + (j)]\n",
-                      taichi_max_num_indices);
+          "};\n";
     }
     if (used.external_ptr) {
       kernel_header +=
           "layout(std430, binding = 4) buffer extr_i32 { int _extr_i32_[]; };\n"
           "layout(std430, binding = 4) buffer extr_f32 { float _extr_f32_[]; "
           "};\n"
           "layout(std430, binding = 4) buffer extr_f64 { double _extr_f64_[]; "
-          "};\n"
-          "#define _ext_ns_i32(x) _extr_i32_[(x) >> 0]\n"
-          "#define _ext_ns_f32(x) _extr_f32_[(x) >> 0]\n"
-          "#define _ext_ns_f64(x) _extr_f64_[(x) >> 0]\n";
+          "};\n";
     }
-    kernel_header +=
-        "#define _Ax_(x) x\n"
-        "#define _At_(x) _Ax_(_at_##x(x))\n";
-    if (used.atomic_float) {  // {{{
-      kernel_header +=
-          "\
-#define _Atmf_Def(Add, _f_, _o_, mem, _32, float) \
-float atomic##Add##_##mem##_f##_32(int addr, float rhs) \
-{ \
-  int old, new, ret; \
-  do { \
-    old = _##mem##_i##_32(addr); \
-    new = floatBitsToInt(_f_(intBitsToFloat(old) _o_ rhs)); \
-  } while (old != atomicCompSwap(_Ax_(_##mem##_i##_32(addr)), old, new)); \
-  return intBitsToFloat(old); \
-}\n\
-#define _Acma_ ,\n\
-#define _Atm_(func, at, x, rhs) _Ax_(func##at(x, rhs))\n\
-_Atmf_Def(Add,, +, mem, 32, float)\n\
-_Atmf_Def(Sub,, -, mem, 32, float)\n\
-_Atmf_Def(Max, max, _Acma_, mem, 32, float)\n\
-_Atmf_Def(Min, min, _Acma_, mem, 32, float)\n\
-"
-#ifdef _GLSL_INT64
-          "\
-_Atmf_Def(Add,, +, 64, double)\n\
-_Atmf_Def(Sub,, -, 64, double)\n\
-_Atmf_Def(Max, max, _Acma_, 64, double)\n\
-_Atmf_Def(Min, min, _Acma_, 64, double)\n\
-"
-#endif
-          "\n";  // discussion:
-                 // https://github.com/taichi-dev/taichi/pull/495#issuecomment-590074123
+    if (used.atomic_float && !opengl_has_GL_NV_shader_atomic_float) {  // {{{
+      kernel_header += STR(float atomicAdd_data_f32(int addr, float rhs) {
+        int old, new, ret;
+        do {
+          old = _data_i32_[addr];
+          new = floatBitsToInt((intBitsToFloat(old) + rhs));
+        } while (old != atomicCompSwap(_data_i32_[addr], old, new));
+        return intBitsToFloat(old);
+      } float atomicSub_data_f32(int addr, float rhs) {
```

Comment:
Thanks for adopting the `STR` macro! Now it looks much cleaner. It seems that clang-format is slightly confused by it and you'll need to turn it on/off as in https://github.com/taichi-dev/taichi/blob/62cdc60f2e3d981b4d5c1c9f60b7590c9e8031a1/taichi/backends/metal/shaders/runtime_kernels.metal.h#L42

so that it will format the contents correctly.

Reply:
So we should guard like:
```
// clang-format off
STR(...)
// clang-format on
```
To make clang-format not formatting my glsl code?

Reply:
Actually, we should turn it on inside `STR`: https://github.com/taichi-dev/taichi/blob/62cdc60f2e3d981b4d5c1c9f60b7590c9e8031a1/taichi/backends/metal/shaders/runtime_kernels.metal.h#L45

Reply:
You probably have already noticed this, but using `STR()` will flatten all your code into one line. So your OpenGl shader source code becomes harder to read.. E.g. this is what i got in Metal:

```
template <typename T, typename G> T union_cast(G g) { static_assert(sizeof(T) == sizeof(G), "Size mismatch"); return *reinterpret_cast<thread const T *>(&g); } inline int ifloordiv(int lhs, int rhs) { const int intm = (lhs / rhs); return (((lhs * rhs < 0) && (rhs * intm != lhs)) ? (intm - 1) : intm); } int32_t pow_i32(int32_t x, int32_t n) { int32_t tmp = x; int32_t ans = 1; while (n) { if (n & 1) ans *= tmp; tmp *= tmp; n >>= 1; } return ans; } float fatomic_fetch_add(device float *dest, const float operand) { bool ok = false; float old_val = 0.0f; while (!ok) { old_val = *dest; float new_val = (old_val + operand); ok = atomic_compare_exchange_weak_explicit( (device atomic_int *)dest, (thread int *)(&old_val), *((thread int *)(&new_val)), metal::memory_order_relaxed, metal::memory_order_relaxed); } return old_val; } float fatomic_fetch_min(device float *dest, const float operand) { bool ok = false; float old_val = 0.0f; while (!ok) { old_val = *dest; float new_val = (old_val < operand) ? old_val : operand; ok = atomic_compare_exchange_weak_explicit( (device atomic_int *)dest, (thread int *)(&old_val), *((thread int *)(&new_val)), metal::memory_order_relaxed, metal::memory_order_relaxed); } return old_val; } float fatomic_fetch_max(device float *dest, const float operand) { bool ok = false; float old_val = 0.0f; while (!ok) { old_val = *dest; float new_val = (old_val > operand) ? old_val : operand; ok = atomic_compare_exchange_weak_explicit( (device atomic_int *)dest, (thread int *)(&old_val), *((thread int *)(&new_val)), metal::memory_order_relaxed, metal::memory_order_relaxed); } return old_val; }
```

Reply:
Oh, I see, but as long as we can compile, this is good. Also we'd better use `/* clang-format on */` inside `STR`.

Reply:
Ack

Reply:
I did a bit more investigation and unfortunately, there doesn't seem to be a way to perfectly resolve this (preserve change-of-line and spaces, while enabling clang-format) within C++.

It is possible to resolve this using a custom command in CMake, but that's a slightly more complex solution. Let's stick to the current solution until some better reason appears to force us to switch to the *CMake + manual preprocessing* approach.

Reply:
Right, i searched for something like keeping the newlines inside macros before, but found no solution :( Like @archibate pointed out, so long as the compiler doesn't complain about the source code, we are good..

Reply:
Sounds good. I'll take a final pass and if there're no other issues this will be merged in.